crates/wasmparser/src/binary_reader.rs

```diff
@@ -242,78 +300,458 @@ impl<'a> BinaryReader<'a> {
+                    .map(|_| Ok((self.read_string()?, self.read_interface_type()?)))
+                    .collect::<Result<_>>()?,
+                result: self.read_interface_type()?,
+            }),
+            0x4b => ComponentTypeDef::Value(self.read_interface_type()?),
+            // Delegate all other values to `read_compound_type`.
+            x => ComponentTypeDef::Compound(self.read_compound_type(x)?),
+        })
+    }
+
     pub(crate) fn read_module_type(&mut self) -> Result<ModuleType<'a>> {
-        let pos = self.original_position();
-        let imports_len = self.read_var_u32()? as usize;
-        if imports_len > MAX_WASM_IMPORTS {
-            return Err(BinaryReaderError::new("imports size is out of bounds", pos));
-        }
-        Ok(ModuleType {
-            imports: (0..imports_len)
-                .map(|_| self.read_import())
-                .collect::<Result<_>>()?,
-            exports: self.read_export_types()?,
+        Ok(match self.read_u8()? {
+            0x01 => ModuleType::Type(self.read_type_def()?),
+            0x02 => ModuleType::Import(self.read_import()?),
+            0x07 => ModuleType::Export {
+                name: self.read_string()?,
+                ty: self.read_type_ref()?,
+            },
+            x => {
+                return Err(BinaryReaderError::new(
+                    format!(
+                        "invalid leading byte (0x{:x}) for module type definition",
+                        x
+                    ),
+                    self.original_position() - 1,
+                ))
+            }
+        })
+    }
+
+    pub(crate) fn read_component_type(&mut self) -> Result<ComponentType<'a>> {
+        Ok(match self.read_u8()? {
+            0x01 => ComponentType::Type(self.read_component_type_def()?),
+            0x02 => ComponentType::Import(self.read_component_import()?),
+            0x07 => ComponentType::Export {
+                name: self.read_string()?,
+                ty: self.read_var_u32()?,
+            },
+            0x09 => {
+                let offset = self.original_position();
+                let alias = self.read_alias()?;
+                match alias {
+                    Alias::OuterType { count, index } => ComponentType::OuterType { count, index },
+                    _ => return Err(BinaryReaderError::new(
+                        "only aliases to outer types are supported in component type definitions",
+                        offset,
+                    )),
+                }
+            }
+            x => {
+                return Err(BinaryReaderError::new(
+                    format!(
+                        "invalid leading byte (0x{:x}) for component type definition",
+                        x
+                    ),
+                    self.original_position() - 1,
+                ))
+            }
         })
     }
 
     pub(crate) fn read_instance_type(&mut self) -> Result<InstanceType<'a>> {
```

Comment:
I think the intention was that components will _only_ import or export items that deal in component types and not core types, so the distinction between the module and component types was needed.

Since a component can only import or export _component_ instances, only a single instance type was needed in the grammar (i.e. that of a component instance). The fact that you can instantiate a module resulting in an entry in the instance index space with a _module type_ is an implementation detail left to the validator/runtime and doesn't need to be expressed in the grammar directly.

When you review the validator you'll see the `InstanceIndexType` enumeration which is used by the validator to track the _actual_ type of the instance in the instance space. If it feels hacky then perhaps this is something we should raise with the spec authors to reconsider this approach.

Reply:
Hmm yeah sort of like how module types don't allow aliases this feels a bit fishy. I'll read more on the validator before drawing any conclusions though